src/main/windows/callsWidgetWindow.ts

```diff
@@ -0,0 +1,197 @@
+import log from 'electron-log';
+
+import {
+    CallsWidgetWindowConfig,
+    CallsWidgetResizeMessage,
+    CallsWidgetShareScreenMessage,
+} from 'types/calls';
+
+import {getLocalPreload} from 'main/utils';
+
+import {
+    MINIMUM_CALLS_WIDGET_WIDTH,
+    MINIMUM_CALLS_WIDGET_HEIGHT,
+    CALLS_PLUGIN_ID,
+} from 'common/utils/constants';
+import Utils from 'common/utils/util';
+import {
+    CALLS_WIDGET_RESIZE,
+    CALLS_WIDGET_SHARE_SCREEN,
+} from 'common/communication';
+
+type LoadURLOpts = {
+    extraHeaders: string;
+}
+
+export default class CallsWidgetWindow extends EventEmitter {
+    public win: BrowserWindow;
+    private main: BrowserWindow;
+    private config: CallsWidgetWindowConfig;
+    private boundsErr: Rectangle = {
+        x: 0,
+        y: 0,
+        width: 0,
+        height: 0,
+    };
+    private offsetsMap = {
+        'calls-widget-menu': {
+            height: 0,
+        },
+    };
+
+    constructor(mainWindow: BrowserWindow, config: CallsWidgetWindowConfig) {
+        super();
+
+        this.config = config;
+        this.main = mainWindow;
+        this.win = new BrowserWindow({
+            width: MINIMUM_CALLS_WIDGET_WIDTH,
+            height: MINIMUM_CALLS_WIDGET_HEIGHT,
+            minWidth: MINIMUM_CALLS_WIDGET_WIDTH,
+            minHeight: MINIMUM_CALLS_WIDGET_HEIGHT,
+            title: 'Calls Widget',
+            fullscreen: false,
+            resizable: false,
+            frame: false,
+            transparent: true,
+            show: false,
+            alwaysOnTop: true,
+            backgroundColor: '#00ffffff',
+            webPreferences: {
+                preload: getLocalPreload('callsWidget.js'),
+            },
+        });
+
+        this.win.once('ready-to-show', () => this.win.show());
+        this.win.once('show', this.onShow);
+        this.win.on('closed', this.onClosed);
+        ipcMain.on(CALLS_WIDGET_RESIZE, this.onResize);
+        ipcMain.on(CALLS_WIDGET_SHARE_SCREEN, this.onShareScreen);
+
+        this.load();
+    }
+
+    public close() {
+        log.debug('CallsWidgetWindow.close');
+        this.win.close();
+    }
+
+    public getServerName() {
+        return this.config.serverName;
+    }
+
+    public getChannelURL() {
+        return this.config.channelURL;
+    }
+
+    private load() {
+        const opts = {} as LoadURLOpts;
+        this.win.loadURL(this.getWidgetURL(), opts).catch((reason) => {
+            log.error(`Calls widget window failed to load: ${reason}`);
+        });
+    }
+
+    private onClosed = () => {
+        log.debug('CallsWidgetWindow.onClosed');
+        this.emit('closed');
+        this.removeAllListeners('closed');
+        ipcMain.off(CALLS_WIDGET_RESIZE, this.onResize);
+        ipcMain.off(CALLS_WIDGET_SHARE_SCREEN, this.onShareScreen);
+    }
+
+    private getWidgetURL() {
+        const u = new url.URL(this.config.siteURL);
+        u.pathname += `/plugins/${CALLS_PLUGIN_ID}/widget/widget.html`;
```

Comment:
I am still considering whether we should be passing this path from the plugin instead of forming it here. I think it should be stable but the flexibility of having it on the server side may help in making future changes backwards compatible.

Reply:
I think this is fine for now. Long term, we will eventually move this widget (or at least the setup for it) into the Desktop App so we won't need this reference. I'd also like to avoid additional message passing stuff if we can.